File: frontend/src/api/leaderboard.ts
import { apiURL } from './config'

export type LeaderboardEntry = {
  rank: number
  nickname: string
  streak: number
}

export type LeaderboardResponse = {
  entries: LeaderboardEntry[]
}

export async function fetchLeaderboard(limit = 50, signal?: AbortSignal): Promise<LeaderboardResponse> {
  const url = `${apiURL('/api/leaderboard')}?limit=${encodeURIComponent(String(limit))}`
  const res = await fetch(url, { method: 'GET', signal })
  if (!res.ok) {
    throw new Error(`leaderboard ${res.status}`)
  }
  const json = (await res.json()) as unknown
  if (!json || typeof json !== 'object' || !('entries' in json)) {
    return { entries: [] }
  }
  const entries = (json as { entries?: unknown }).entries
  return {
    entries: Array.isArray(entries) ? (entries as LeaderboardResponse['entries']) : [],
  }
}

File: backend/internal/handler/leaderboard.go
package handler

import (
	"database/sql"
	"encoding/json"
	"log"
	"net/http"
	"strconv"

	"github.com/KoiralaSam/Mindcare/backend/internal/models/user"
)

// LeaderboardEntryJSON is a public row for streak rankings (no email).
type LeaderboardEntryJSON struct {
	Rank     int    `json:"rank"`
	Nickname string `json:"nickname"`
	Streak   int    `json:"streak"`
}

// LeaderboardResponse is returned by GET /api/leaderboard.
type LeaderboardResponse struct {
	Entries []LeaderboardEntryJSON `json:"entries"`
}

// LeaderboardHandler returns users ordered by streak (highest first).
func LeaderboardHandler(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", http.MethodGet)
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}

		limit := 50
		if q := r.URL.Query().Get("limit"); q != "" {
			n, err := strconv.Atoi(q)
			if err != nil || n < 1 {
				http.Error(w, "invalid limit", http.StatusBadRequest)
				return
			}
			if n > 100 {
				n = 100
			}
			limit = n
		}

		rows, err := user.ListStreakLeaderboard(db, limit)
		if err != nil {
			log.Printf("leaderboard: %v", err)
			http.Error(w, "internal server error", http.StatusInternalServerError)
			return
		}

		out := make([]LeaderboardEntryJSON, len(rows))
		for i := range rows {
			out[i] = LeaderboardEntryJSON{
				Rank:     i + 1,
				Nickname: rows[i].Nickname,
				Streak:   rows[i].Streak,
			}
		}

		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(LeaderboardResponse{Entries: out})
	}
}
